### ISSUE TITLE
(NETPLAY) Clients cause disconnection of all peers when load savestates

### ISSUE THREAD

**Issue by furiadeoso** [libretro/RetroArch#7767]
This issue is happening only when MORE THAN TWO PLAYERS are connected.

If the netplay game have three or more players connected, any client that load a saved state cause disconection of ALL the peers (clients).

This is happening in normal p2p mode, not MITM.

### Expected behavior

Any client should be able to load and extended a savestate. Or at least not cause a total disconnection of all the peers.

### Possible related issues
In our experience doing a lot of netplay gamig we have noticed that when three or more players are involved there are possibilities to get disconnection or desync. At the beginnning we though this could be more related to the game/core, but experience and observation make us think that could be some little bad with three or more players.
This old reported issues could be related:
https://github.com/libretro/RetroArch/issues/6986
https://github.com/libretro/RetroArch/issues/6309

We can test  whatever is necessary to help.

**Comment by bparker06** [comment]
paging @GregorR

**Comment by iosadosd-quarto** [comment]
This happens to me too, and until I stumbled on this thread I couldn't understand why I was getting (seemingly) random disconnections when playing with more than 1 client.

**Comment by furiadeoso** [comment]
@iosadosd-quarto Are you getting random (apparently) disconnections also like us?. It is mean,  not only when a client press load state, but also in other moments but only with more than 1 clients.

**Comment by iosadosd-quarto** [comment]
Yes, mostly in "other", random moments: clients who I've played with have never tried to (willingly) load a save - not that I know of. I assume that those random disconnections are when RetroArch tries to sync savestates when there's lag or some instants of slow connection.
In fact, I've never had random client disconnections when playing with people geographically close to me or with very stable connections - or maybe it's because we happened to play some arcade games which don't support savestates, I've never looked too much into it until I read this thread.

**Comment by furiadeoso** [comment]
Try disabling check-frames (0). The check moment, in my experience, can cause problems when there are more than 1 client.
In fact there is a proved bug: https://github.com/libretro/RetroArch/issues/6986
But in my opinion could be more not proved related to 2 or more clients and syncing.

**Comment by iosadosd-quarto** [comment]
@furiadeoso I was about to replay that I've never used runahead to reduce latency, but then I realised that netplay_check_frames is not a Latency setting, but a Netplay related setting, which is not shown in the front-end unless I turn on "Show advanced settings". I'll try this out the next time I'll host a match for more than 2 players and I'll let you know.
Just to be sure: should I have the clients set check frames to 0, or is it just for the host?

**Comment by furiadeoso** [comment]
> Just to be sure: should I have the clients set check frames to 0, or is it just for the host?
Just for the host.
> 
> EDIT: Also, how do you make sure that the host and clients are in sync if checking every X frames is in fact disabled? Wouldn't this put the netplay session at risk of desynchronization?

You can ask your mates for the the actual score.

**Comment by furiadeoso** [comment]
> Right  and if/when we find out that we are desynced, how do we roll back if loading a state will cause disconnections? Do we have to start over every time?

It's easy. The host presses save and load state quickly (1 and 3 keys I think) to extend his/her state.
When the issue is resolved clients will be able too.

**Comment by iosadosd-quarto** [comment]
Oh ok so hosts **can** load a savestate without causing disconnection of clients! For some reason I thought neither the clients nor the host could 😅

**Comment by GregorR** [comment]
Can't repro. The "correct" (or at least expected) behavior is: If there are two players, then either can load state; if there are three or more players, only the host can load state, and if a client attempts to load state, they will be disconnected. This is the only behavior I'm capable of reproducing.

In trying to repro this, I uncovered a bug regarding loading savestates poorly timed with respect to pausing, which *does* cause the disconnection of all clients (for a very bizarre reason, as it's actually the *clients* disconnecting, not the server dropping them). Perhaps it is related.

**Comment by furiadeoso** [comment]
> Can't repro. The "correct" (or at least expected) behavior is: If there are two players, then either can load state; if there are three or more players, only the host can load state, and if a client attempts to load state, the client in question will be disconnected. This is the only behavior I'm capable of reproducing.
> 
> In trying to repro this, I uncovered a bug regarding loading savestates poorly timed with respect to pausing, which _does_ cause the disconnection of all clients (for a very bizarre reason, as it's actually the _clients_ disconnecting, not the server dropping them). Perhaps it is related.

It is true. I am trying to reproduce it now and only the client that is loading drops. I guess I was wrong when  did test. I will correct the first post.

When we use the pause we have many disconnection problems. I'm going to leave the issue open so you can refer to it when you review the bug you have mentioned.
